sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -445,23 +643,10 @@ cdef class DistanceMetric{{name_suffix}}:
             The shape (Nx, Ny) array of pairwise distances between points in
             X and Y.
         """
-        cdef cnp.ndarray[{{DTYPE_t}}, ndim=2, mode='c'] Xarr
-        cdef cnp.ndarray[{{DTYPE_t}}, ndim=2, mode='c'] Yarr
-        cdef cnp.ndarray[{{DTYPE_t}}, ndim=2, mode='c'] Darr
-
-        Xarr = np.asarray(X, dtype={{DTYPE}}, order='C')
-        self._validate_data(Xarr)
-        if Y is None:
-            Darr = np.zeros((Xarr.shape[0], Xarr.shape[0]),
-                         dtype={{DTYPE}}, order='C')
-            self.pdist(Xarr, Darr)
-        else:
-            Yarr = np.asarray(Y, dtype={{DTYPE}}, order='C')
-            self._validate_data(Yarr)
-            Darr = np.zeros((Xarr.shape[0], Yarr.shape[0]),
-                         dtype={{DTYPE}}, order='C')
-            self.cdist(Xarr, Yarr, Darr)
-        return Darr
+        if not issparse(X) and (Y is None or not issparse(Y)):
+            return self._pairwise_dense(X, Y)
+        if issparse(X) and (Y is None or issparse(Y)):
+            return self._pairwise_sparse(X, Y)
```

Comment:
Actually it would be nicer to make this support the sparse x dense case by doing the trick to cast the dense array as a CSR using the tiled `np.arange()` indices.